File: 871-minimum-number-of-refueling-stops/871-minimum-number-of-refueling-stops.py
class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        
        # if target is less than current capacity
        if target <= startFuel: return 0
            
        dp = [startFuel] + [0] * len(stations)        
        for i, (curr_distance, fuel) in enumerate(stations):
            for j in range(i, -1, -1):
                if dp[j] >= curr_distance:
                    dp[j +1] = max(dp[j + 1], dp[j] + fuel)

        for idx, max_reach in enumerate(dp):
            if max_reach >= target: return idx
        return -1

File: 871-minimum-number-of-refueling-stops/871-minimum-number-of-refueling-stops.cpp
// typedef unsigned long long ;
class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        vector<unsigned long long>dp(stations.size() + 1, 0);
        dp[0] = startFuel;
        
        for(int i = 0; i < stations.size(); i ++) {
            // const auto [cur_distance, fuel] = stations[i];
            for (int j = i; j >-1; j--) {
                if (dp[j] >= stations[i][0]) {
                    dp[j + 1] = max(dp[j + 1], dp[j] + stations[i][1]);
                }
            }
        }
        
        for (int i = 0; i < dp.size(); i ++) {
            if (dp[i] >= target) return i;
        }
        return -1;
    }
};